File: bonsai_adapter_ros_clf/src/main/java/de/unibi/citec/clf/bonsai/ros/actuators/RosHandOverActuator.java
package de.unibi.citec.clf.bonsai.ros.actuators;

import actionlib_msgs.GoalStatus;
import actionlib_msgs.GoalStatusArray;
import com.github.rosjava_actionlib.ActionClient;
import com.github.rosjava_actionlib.ActionClientListener;
import de.unibi.citec.clf.bonsai.actuators.HandOverActuator;
import de.unibi.citec.clf.bonsai.core.configuration.IObjectConfigurator;
import de.unibi.citec.clf.bonsai.core.exception.ConfigurationException;
import org.ros.namespace.GraphName;
import org.ros.node.ConnectedNode;
import de.unibi.citec.clf.bonsai.ros.RosNode;
import hand_over_msgs.HandOverActionFeedback;
import hand_over_msgs.HandOverActionGoal;
import hand_over_msgs.HandOverActionResult;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.Future;
import org.ros.message.Duration;

/**
 * Created by lruegeme
 */
public class RosHandOverActuator extends RosNode implements HandOverActuator {

    String serverTopic;
    private GraphName nodeName;

    private ActionClient<HandOverActionGoal, HandOverActionFeedback, HandOverActionResult> ac = null;
    private org.apache.log4j.Logger logger = org.apache.log4j.Logger.getLogger(getClass());

    public RosHandOverActuator(GraphName gn) {
        initialized = false;
        this.nodeName = gn;
    }

    @Override
    public void configure(IObjectConfigurator conf) throws ConfigurationException {
        this.serverTopic = conf.requestValue("topic");
    }

    @Override
    public GraphName getDefaultNodeName() {
        return nodeName;
    }

    @Override
    public void onStart(final ConnectedNode connectedNode) {
        ac = new ActionClient(connectedNode, this.serverTopic, HandOverActionGoal._TYPE, HandOverActionFeedback._TYPE, HandOverActionResult._TYPE);
        if (ac.waitForActionServerToStart(new Duration(2.0))) {
            initialized = true;
            logger.debug("RosHandOverActuator started");
        } else {
            logger.debug("RosHandOverActuatortimeout after 2sec " + this.serverTopic);
        }

    }

    @Override
    public void destroyNode() {
        if(ac!=null) ac.finish();
    }



    @Override
    public Future<Boolean> handOver(String group_name, byte type) throws IOException {
        if (!ac.waitForActionServerToStart(new Duration(1))) {
            logger.error("action server not started on " + this.serverTopic);
        }

        HandOverActionGoal goalMessage = ac.newGoalMessage();
        goalMessage.getGoal().setGroupName(group_name);
        goalMessage.getGoal().setType(type);
        logger.info("sending hand_over goal for group " + group_name + " with type " + type + " and id:" + goalMessage.getGoalId());
        return ac.sendGoal(goalMessage).toBooleanFuture();
    }



}

File: bonsai_adapter_ros_clf/src/main/java/de/unibi/citec/clf/bonsai/ros/actuators/RosHandOverActuator.kt
package de.unibi.citec.clf.bonsai.ros.actuators

import com.github.rosjava_actionlib.ActionClient
import de.unibi.citec.clf.bonsai.actuators.HandOverActuator
import de.unibi.citec.clf.bonsai.core.configuration.IObjectConfigurator
import de.unibi.citec.clf.bonsai.core.exception.ConfigurationException
import de.unibi.citec.clf.bonsai.ros.RosNode
import hand_over_msgs.*
import org.ros.message.Duration
import org.ros.namespace.GraphName
import org.ros.node.ConnectedNode
import java.io.IOException
import java.util.concurrent.Future

/**
 * Created by lruegeme
 */
class RosHandOverActuator(private val nodeName: GraphName) : RosNode(), HandOverActuator {

    private lateinit var handoverTopic: String
    private lateinit var measureTopic: String

    private var acHandover: ActionClient<HandOverActionGoal, HandOverActionFeedback, HandOverActionResult>? = null
    private var acMeasure: ActionClient<MeasureForceActionGoal, MeasureForceActionFeedback, MeasureForceActionResult>? =
        null
    private val logger = org.apache.log4j.Logger.getLogger(javaClass)

    init {
        initialized = false
    }

    @Throws(ConfigurationException::class)
    override fun configure(conf: IObjectConfigurator) {
        this.handoverTopic = conf.requestValue("handover_topic")
        this.measureTopic = conf.requestValue("measure_topic")
    }

    override fun getDefaultNodeName(): GraphName {
        return nodeName
    }

    override fun onStart(connectedNode: ConnectedNode) {
        acHandover = ActionClient(
            connectedNode,
            this.handoverTopic,
            HandOverActionGoal._TYPE,
            HandOverActionFeedback._TYPE,
            HandOverActionResult._TYPE
        )
        acMeasure = ActionClient(
            connectedNode,
            this.measureTopic,
            MeasureForceActionGoal._TYPE,
            MeasureForceActionFeedback._TYPE,
            MeasureForceActionResult._TYPE
        )

        var c = 0;

        if (acHandover?.waitForActionServerToStart(Duration(20.0)) == true) {
            c++
        } else {
            logger.error("could not connect to $handoverTopic")
        }

        if (acMeasure?.waitForActionServerToStart(Duration(20.0)) == true) {
            c++
        } else {
            logger.error("could not connect to $measureTopic")
        }

        if(c == 2) {
            initialized = true
            logger.info("${javaClass.simpleName} started")
        }

    }

    override fun destroyNode() {
        acHandover?.let { it.finish() }
        acMeasure?.let { it.finish() }
    }


    @Throws(IOException::class)
    override fun handOver(group_name: String, type: Byte): Future<Boolean> {
        acHandover?.let { client ->
            val goalMessage = client.newGoalMessage()
            goalMessage.goal.groupName = group_name
            goalMessage.goal.type = type
            logger.info("sending hand_over goal for group $group_name with type $type and id: ${goalMessage.goalId}")
            return client.sendGoal(goalMessage).toBooleanFuture()
        }

        throw IOException("action server failure ${this.handoverTopic}")
    }

    @Throws(IOException::class)
    override fun checkForce(group_name: String, threshold: Float): Future<Boolean>? {
        acMeasure?.let { client ->
            val goalMessage = client.newGoalMessage()
            //goalMessage.goal.groupName = group_name
            goalMessage.goal.threshold = threshold
            logger.info("sending measure goal with threshold $threshold and id: ${goalMessage.goalId}")
            return client.sendGoal(goalMessage).toBooleanFuture()
        }

        throw IOException("action server failure ${this.measureTopic}")
    }


}
